lib/types.nix

```diff
@@ -469,7 +469,7 @@ rec {
       descriptionClass = "noun";
       check = x: isDerivation x || isStorePath x;
       merge = loc: defs:
-        let res = mergeOneOption loc defs;
+        let res = mergeEqualOption loc defs;
```

Comment:
This seems risky, because equality on derivation-like things is not accurate. We don't want to pick an arbitrary package when that happens.

```
> zlib == zlib // { yikes = true; }
true
```

Reply:
Hmm that's a good point. I think we can compare on the `outPath` then?

The issue that we're trying to avoid here is that the same package can be added multiple times with the list syntax, and we don't want to get an error from the module system in that case when they are exactly the same, in that case we can just deduplicate and continue.

Reply:
The does mean that things like `passthru` are not taken into account in the deduplication though.

Reply:
> That would mean that things like passthru are not taken into account in the deduplication though.

This could be a problem: I know several packages that provide plugins this way and you may want to override the package to add your own plugin, for example.

Reply:
But the final package with the plugin added would still have a different `outPath` then, wouldn't it?

The passthru itself is not part of the drv, so it can only provide things at eval time, but it cannot affect the resulting build output.

Reply:
I don't think it's necessary for this change to be part of this PR.

Final thought: probably an appropriate equality for packages depends on the use case, so we might want to do this not in `types.nix`, but rather make exceptions where in concrete cases where we can reason about which definitions are equivalent given how the option is used.

Reply:
See https://github.com/NixOS/nixpkgs/pull/223064